packages/flutter/test/gestures/events_test.dart

```diff
@@ -110,4 +112,342 @@ void main() {
       expect(event.buttons, kPrimaryButton);
     });
   });
+
+  test('paintTransformToPointerEventTransform', () {
+    Matrix4 original = Matrix4.identity();
+    Matrix4 changed = PointerEvent.paintTransformToPointerEventTransform(original);
+    expect(changed, original);
+
+    original = Matrix4.identity()..scale(3.0);
+    changed = PointerEvent.paintTransformToPointerEventTransform(original);
+    expect(changed, isNot(original));
+    original
+      ..setColumn(2, Vector4(0, 0, 1, 0))
+      ..setRow(2, Vector4(0, 0, 1, 0));
+    expect(changed, original);
+  });
+
+  test('transformPosition', () {
+    const Offset position = Offset(20, 30);
+    expect(PointerEvent.transformPosition(null, position), position);
+    expect(PointerEvent.transformPosition(Matrix4.identity(), position), position);
+    final Matrix4 transform = Matrix4.translationValues(10, 20, 0);
+    expect(PointerEvent.transformPosition(transform, position), const Offset(20.0 + 10.0, 30.0 + 20.0));
+  });
```

Comment:
Having this test with a transformation containing an out of plane rotation (rotateX or rotateY) would be useful as that is a less obvious case that has a bug in the Transform widget.

Reply:
There's the complex "Perspective transform on scrollable" test at the very bottom of the PR that rotates around X and verifies that scrolling still works.

Reply:
I hope it's okay, but created the test that I proposed to show that it fails: https://github.com/goderbauer/flutter/compare/s2...spkersten:goderbauer-s2-rotation-test?expand=1

Reply:
I've looked at the "Perspective transform on scrollable" test and it made me really confused. It also fails when I removed the perspective (`setEntry(3,2, 0.001)`) and only keep the rotation. So I add the widget tree in an app to see what happens:
```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(home: Foo());
}

class Foo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Transform(
      transform: Matrix4.identity()
        ..setEntry(3, 2, 0.001)
        ..rotateX(math.pi / 4)
      ,
      child: Container(
        padding: EdgeInsets.all(16),
        color: Colors.yellow,
        child: Center(
          child: Container(
            width: 200,
            child: ListView.builder(
//            controller: controller,
              cacheExtent: 0.0,
              itemBuilder: (BuildContext context, int index) {
                return Container(
                  height: 100.0,
                  color: index % 2 == 0 ? Colors.blue : Colors.red,
                  child: Text('Tile $index'),
                );
              },
            ),
          ),
        ),
      ),
    );
  }
}
```
I've added the `DecoratedBox` to see more clearly what is going on. Here is a screenshot:
<img width="581" alt="Screenshot 2019-05-24 at 19 42 30" src="https://user-images.githubusercontent.com/4456832/58346827-651cfb80-7e5c-11e9-8e5c-61171c3fc31e.png">
So it looks like the children aren't transformed at all!

Reply:
> https://github.com/goderbauer/flutter/compare/s2...spkersten:goderbauer-s2-rotation-test?expand=1

Yes, applying `MatrixUtils.transformPoint` in both directions will not work as that method completely drops the z value on the floor as you already found out. If you use that method to transform a point P(x, y, 0) on the xy-plane of a x-rotated render plane you will end up with a point P'(x', y', z') and z' will most likely not be 0. If you now take the x' and y' of that point and run it through `transformPoint` again (which just sets z to 0), you will not end up with x and y again.

In combination with `PointerEvent.removePerspectiveTransform` it works because that method modifies the transform matrix in a way to map point P(x, y, 0) from the xy-plane of the render plane to the point P'(x', y', 0) on the xy-plane of the screen (that's the point on the screen at which P will be drawn). Because z is zero for both points, we can now use `MatrixUtils.transformPoint` to correctly transform between the two.

Hope this makes sense? Looking at the other example you posted now...

Reply:
About your second comment: That looks like an iOS-specific rendering bug. On Android your example renders fine. Let me do a little investigation and file a separate bug about it.

Reply:
Actually, it looks like it was a bug with the old engine version that this PR was synced to. I just pulled in master and everything renders correctly. (And hit testing also works as expected.)